File: js/IsPointInPath.js
function IsPointInPath(num, ctx) {
    var pointsPerPath = 10;

    for(i = 0; i < num / 10; i++) {

    var x1 = Math.floor(Math.random() * canvas.width / 2);
    var x2 = Math.floor(Math.random() * canvas.width / 2);
    var y1 = Math.floor(Math.random() * canvas.height / 2);
    var y2 = Math.floor(Math.random() * canvas.height / 2);

    ctx.strokeStyle="blue";
    ctx.beginPath();
    ctx.rect(x1, y1, x2, y2);
    ctx.stroke();

    for (j = 0; j < pointsPerPath; j++) {
        var x = Math.floor(Math.random() * canvas.width);
        var y = Math.floor(Math.random() * canvas.height);
        
        var col;
        if (ctx.isPointInPath(x, y)) {
            col = "red";
        } else {
            col = "green";
        }

        ctx.beginPath();
        ctx.fillStyle = col;
        ctx.arc(x, y, 5, 0, 2 * Math.PI, false);
        ctx.closePath();
        ctx.fill();
    } 
    }
}

File: hs/IsPointInPath.hs
{-# LANGUAGE ApplicativeDo     #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE OverloadedStrings #-}
module IsPointInPath (benchmark, summary) where

import           Control.Monad.Compat

import           Data.Foldable        (for_, sequenceA_)

import           Graphics.Blank
import           Prelude.Compat
import           Utils

benchmark :: CanvasBenchmark
benchmark ctx = for_ [1..rounds] (const (internal ctx))

internal :: CanvasBenchmark
internal ctx = do
    pathX1 <- randomXCoord ctx
    pathX2 <- randomXCoord ctx
    pathY1 <- randomYCoord ctx
    pathY2 <- randomYCoord ctx
    points <- replicateM pointsPerPath $ (,) <$> randomXCoord ctx <*> randomYCoord ctx
    send' ctx $ sequenceA_ [ isInPath (pathX1, pathX2, pathY1, pathY2) points ]

summary :: String
summary = "IsPointInPath"

rounds :: Int
rounds = 100

pointsPerPath :: Int
pointsPerPath = 10

pointRadius :: Double
pointRadius = 5

isInPath :: Path -> [Point] -> Canvas ()
isInPath (pathX, pathY, pathW, pathH) points = do
    strokeStyle("blue");
    beginPath();
    rect(pathX, pathY, pathW, pathH);
    cmds <- sequence [ do b <- isPointInPath(x, y);
                          return $ do
                              beginPath();
                              fillStyle(if b then "red" else "green");
                              arc(x, y, pointRadius, 0, pi*2, False);
                              fill();
                     | (x, y) <- points
                     ]
    stroke();
    sequence_ cmds
